### ISSUE TITLE
Decrypting frc cookie - Help wanted from java users

### ISSUE THREAD

**Issue by mkb79** [mkb79/Audible#42]
I have this [java script](https://github.com/mkb79/Audible/files/6151259/encryptionhelper.txt) which encrypts a strigified JSON object. 

Amazon apps using this string as the "frc" cookie. My app creates the frc cookie from custom bytes. Amazon still accepts this frc cookie. But I will be prepared for the future and build the frc cookie the same way as Amazon does. 

I've tried to rebuild the java script to python

```python
import binascii
import base64
import hmac
import hashlib
import gzip

from Crypto.Cipher import AES


def check_sig(pw, iv, data, sig):
    key = get_key_from_password(pw, b"HmacSHA256")
    x = hmac.new(key, iv + data, hashlib.sha256).digest()
    x = x[:len(sig)]
    assert sig == x


def decrypt_frc(pw, iv, data):
    key = get_key_from_password(pw, b"AES/CBC/PKCS7Padding")
    crypter = AES.new(key, AES.MODE_CBC, iv)
    decrypted_data = crypter.decrypt(data)
    return decrypted_data


def split_frc(frc):
    frc += (len(frc) % 4) * '='
    frc = base64.b64decode(frc)

    sig = frc[1:9]
    iv = frc[9:25]
    data = frc[25:]

    return sig, iv, data


def get_key_from_password(pw, salt):
    salt = "HmacSHA256"
    return hashlib.pbkdf2_hmac('sha1', pw, salt, 1000, 16)


if __name__ == '__main__':
    frc_cookie = ''
    pw = 'MUST_FIND_IT_OUT'

    sig, iv, data = split_frc(frc)
    check_sig(pw, iv, data, sig)
    decrypted = decrypt_frc(key, iv, data)
    decompressed = gzip.decompress(decrypted)
```

Can someone with JAVA and Python knowledge take a look on my code if I have take a misstake?

**Comment by mkb79** [comment]
Okay, I've found out how to decrypt the frc cookie myself. The 'pw' is the device serial.

**Comment by mkb79** [comment]
Here is an example of the JSON data from a decrypted frc cookie

```json
{
    "ApplicationVersion": "3.35.1",
    "DeviceOSVersion": "iOS/14.1",
    "ScreenWidthPixels": "375",
    "TimeZone": "+02:00",
    "ScreenHeightPixels": "812",
    "ApplicationName": "Audible",
    "DeviceJailbroken": false,
    "DeviceLanguage": "de-DE",
    "DeviceFingerprintTimestamp": "...",
    "ThirdPartyDeviceId": "...",
    "DeviceName": "iPhone",
    "Carrier": "..."
}
```

**Comment by mkb79** [comment]
Here is my optimized code to encrypt and decrypt the frc cookie:

```python
import base64
import gzip
import hashlib
import hmac
import json
import secrets
from io import BytesIO
from functools import partialmethod
from typing import Tuple, Union

from pyaes import AESModeOfOperationCBC, Encrypter, Decrypter


FRC_SIG_SALT: bytes = b'HmacSHA256'
FRC_AES_SALT: bytes = b'AES/CBC/PKCS7Padding'


class FrcCookieHelper:
    def __init__(self, password: str) -> None:
        self.password = password.encode()

    def _get_key(self, salt: bytes) -> bytes:
        return hashlib.pbkdf2_hmac('sha1', self.password, salt, 1000, 16)

    get_signature_key = partialmethod(_get_key, FRC_SIG_SALT)

    get_aes_key = partialmethod(_get_key, FRC_AES_SALT)

    @staticmethod
    def unpack(frc: str) -> Tuple[bytes, bytes, bytes]:
        pad = len(frc) % 4 * '='
        frc = BytesIO(base64.b64decode(frc+pad))
        frc.seek(1)  # the first byte is always 0, skip them
        return frc.read(8), frc.read(16), frc.read()  # sig, iv, data

    @staticmethod
    def pack(sig: bytes, iv: bytes, data: bytes) -> str:
        frc = b'\x00' + sig[:8] + iv[:16] + data
        frc = base64.b64encode(frc).strip(b'=')
        return frc.decode()

    def verify_signature(self, frc: str) -> bool:
        key = self.get_signature_key()
        sig, iv, data = self.unpack(frc)
        new_signature = hmac.new(key, iv + data, hashlib.sha256).digest()
        new_signature = new_signature[:len(sig)]
        return sig == new_signature

    def decrypt(self, frc: str, verify_signature: bool = True) -> str:
        if verify_signature:
            self.verify_signature(frc)

        key = self.get_aes_key()
        sig, iv, data = self.unpack(frc)
        decrypter = Decrypter(AESModeOfOperationCBC(key, iv))
        decrypted = decrypter.feed(data) + decrypter.feed()
        decompressed = gzip.decompress(decrypted)
        return decompressed

    def encrypt(self, data: Union[str, dict]) -> str:
        if isinstance(data, dict):
            data = json.dumps(data, indent=2, separators=(',', ' : ')).encode()

        compressed = BytesIO()
        with gzip.GzipFile(fileobj=compressed, mode='wb', mtime=False) as f:
            f.write(data)
        compressed.seek(8)
        compressed.write(b'\x00\x13')
        compressed = compressed.getvalue()
        
        key = self.get_aes_key()
        iv = secrets.token_bytes(16)

        encrypter = Encrypter(AESModeOfOperationCBC(key, iv))
        encrypted = encrypter.feed(compressed) + encrypter.feed()

        key = self.get_signature_key()
        signature = hmac.new(key, iv + encrypted, hashlib.sha256).digest()

        return self.pack(signature, iv, encrypted)
```

**Comment by eode** [comment]
Hey, I don't have the knowledge of Amazon's code to verify or compare it, or to evaluate this -- but:

If both of these are true:
  * amazon updates their function for authentication enough that keeping python synchronized is a pain
  * the function that they use is atomic in javascript (in the sense that it can be used as a regular function call)
Or if this is:
  * You want to test your function's output to verify it matches Amazon's output

You could use something like `pyv8` or `bond` or (if you're using qt/webkit) `evaluateJavaScript()`.

In any case, if you can get the native function to be callable from Python, then you'd have a pretty foolproof way of ensuring that your results match theirs.

Note:  Inter-language frameworks can be a rabbit-hole.  This isn't a request for you to use the options mentioned, I just thought this info might be interesting or useful.

**Comment by mkb79** [comment]
Hi, thank you for your hint.

Life would much easier using Python and Java Script in one project.

But:
I wrote my Python apps with Pythonista for iOS. That's the reason why I can't use Java Script in my Audible projects. And that's the same reason why I use native Python packages only as dependencies (or such package which are shipped with Pythonista).

Have Amazon changed encrypting the frc cookies?

**Comment by figamore** [comment]
> Hi, thank you for your hint.
> 
> Life would much easier using Python and Java Script in one project.
> 
> But:
> I wrote my Python apps with Pythonista for iOS. That's the reason why I can't use Java Script in my Audible projects. And that's the same reason why I use native Python packages only as dependencies (or such package which are shipped with Pythonista).
> 
> Have Amazon changed encrypting the frc cookies?

Did you ever figure out how to decrypt/encrypt the frc cookies?

They seem to be outputting an frc string containing "\/" I tried your code but it only produces "/" in the string. Any clue why that is?

**Comment by mkb79** [comment]
> Did an you ever figure out how to decrypt/encrypt the frc cookies?

[Here](https://github.com/mkb79/Audible/issues/42#issuecomment-801905044) is the working code to decrypt/encrypt a frc cookie. As far as I found out, the password is the device serial.

> They seem to be outputting an frc string containing "/" I tried your code but it only produces "/" in the string. Any clue why that is?

What you mean with this? Have you an example?

**Comment by mkb79** [comment]
@figamore 

[Here](https://gist.github.com/mkb79/2a3f0acb3a75d3cf82d0c85a29497d22) you can find a gist to register a Goodreads device using a frc cookie.

**Comment by seidnerj** [comment]
This is awesome work! Great job! Where did you get the original Java implementation from? I am looking at some other Amazon constructs and am looking for some hints.

**Comment by seidnerj** [comment]
Specifically there's a cookie called CES that I suspect has something to do with captcha authentication.

If it is not present/incorrect, even if the captcha is entered correctly you get an error message that's identical to the one you get when you enter an incorrect captcha.

**Comment by seidnerj** [comment]
Sorry, I meant it is a field in the login form, not a cookie. It is in the Amazon shopping app/mobile web and only appears during a captcha challenge.

It is indeed similar to the CSRF in the sense that it is sent from the server and posted back but for some reason if you try to login programmatically it seems that the value for CRS sent from the server and then reposted is incorrect.

**Comment by mkb79** [comment]
I doesn’t know anything about the „ape:“ prefix. But it from the name I would think first on „Authenticated Permutation-Based Encryption for Lightweight Cryptography (APE)“?! But I‘m not sure.

**Comment by seidnerj** [comment]
I looked it up, maybe - either way I haven't been able to make progress there for time being 🤷‍♂️

**Comment by mkb79** [comment]
I would also like to take a closer look. Can I also use the Amazon iOS Shopping App for this? Or does it work differently there?

**Comment by seidnerj** [comment]
Yes, thanks a lot! You'd need to get it to issue you with a captcha challenge though, otherwise this field isn't going to be present. I think an easier way is to get this through the mobile web version of amazon.com. For the most part, the Amazon app is simply a wrapper for the mobile website.

**Comment by mkb79** [comment]
Okay. Now I have to figure out how to force a captcha prompt. I‘m think I must disable 2FA?!

**Comment by seidnerj** [comment]
If you start a session with no cookies at all, then type the wrong password a couple of times, you usually get captcha 🙂

**Comment by seidnerj** [comment]
I'm not sure about having to disable 2FA, in my case it sometimes sends an email with an OTP, so I guess the 2FA is enabled but I still get a captcha if I do what I suggested in my previous post.

**Comment by seidnerj** [comment]
> Hey, I don't have the knowledge of Amazon's code to verify or compare it, or to evaluate this -- but:
> 
> 
> 
> If both of these are true:
> 
>   * amazon updates their function for authentication enough that keeping python synchronized is a pain
> 
>   * the function that they use is atomic in javascript (in the sense that it can be used as a regular function call)
> 
> 
> 
> Or if this is:
> 
>   * You want to test your function's output to verify it matches Amazon's output
> 
> 
> 
> You could use something like `pyv8` or `bond` or (if you're using qt/webkit) `evaluateJavaScript()`.
> 
> 
> 
> In any case, if you can get the native function to be callable from Python, then you'd have a pretty foolproof way of ensuring that your results match theirs.
> 
> 
> 
> Note:  Inter-language frameworks can be a rabbit-hole.  This isn't a request for you to use the options mentioned, I just thought this info might be interesting or useful.

Pyv8 is no longer maintained, luckily it was forked by Cloudare and is still being actively maintained. Under project name "Stpyv8", see here: https://github.com/cloudflare/stpyv8

**Comment by seidnerj** [comment]
Yes, there are more form values that appear with this same prefix, thing is, in all other cases (such as the ones you've listed below), the data after that prefix its simply base64 encoded and when you decode it you get plain text that's not encrypted. For example, for the values you provided:

`aHR0cHM6Ly93d3cuYW1hem9uLmNvbS8/X2VuY29kaW5nPVVURjgmcmVmXz1uYXZtX2hkcl9zaWduaW4=` decodes to "https://www.amazon.com/?_encoding=UTF8&ref_=navm_hdr_signin"

Whereas `NkVFS0RZNk1BS0pNMTY4MVg0NlA=`decodes to "6EEKDY6MAKJM1681X46P", which if you look closely, is essentially the previous request ID.

In both cases, the "payload" is simply base64 encoded but not encrypted. In the case of the CES value, it seems the data is encrypted prior to the base64 encoding.

**Comment by mkb79** [comment]
Yeah, in Python the code looks:

```Python 
def build_client_id(serial: str) -> str:
    client_id = serial.encode() + b"#A2CZJZGLK2JJVM"  # iPhone device type
    return client_id.hex()
```
